File: DSA/src/Questions/StackQueue/CalculateExpression.java
package Questions.StackQueue;

import java.util.Stack;

public class CalculateExpression {
    public long resolve(long a, long b, char Operator) {
        if (Operator == '+')
            return a + b;
        else if (Operator == '-')
            return a - b;
        else if (Operator == '*')
            return a * b;
        else
            return a / b;
    }

    public int evalRPN(String[] tokens) {
        Stack<Long> s = new Stack<>();
        for (String token : tokens) {
            if (token.length() == 1 && token.charAt(0) < '0') {
                long integer2 = s.pop();
                long integer1 = s.pop();
                char operator = token.charAt(0);
                long result = resolve(integer1, integer2, operator);
                // System.out.println(result);
                s.push(result);
            } else {
                s.push(Long.parseLong(token));
            }
        }
        return s.pop().intValue();
    }
    public static void main(String[] args) {
        CalculateExpression calculateExpression = new CalculateExpression();
        String[] tokens = {"2", "1", "+", "3", "*"};
        int result = calculateExpression.evalRPN(tokens);
        System.out.println("Result: " + result);
    }
}

File: DSA/Python/Questions/StackQueue/CalculateExpression.py
"""
# Question: Evaluate Reverse Polish Notation
# Link: https://leetcode.com/problems/evaluate-reverse-polish-notation/

# Calculate result of expression in Reverse Polish Notation

# Time Complexity: O(n)
# Space Complexity: O(n)

# Algorithm:
# 1. Process tokens one by one
# 2. Push numbers onto stack
# 3. Evaluate operators with top two numbers
# 4. Return final result

# Key Components:
# - eval_rpn(): Main expression evaluator
# - resolve(): Helper for operator evaluation
# - Stack-based calculation
"""


class CalculateExpression:
    def resolve(self, a: int, b: int, operator: str) -> int:
        if operator == "+":
            return a + b
        elif operator == "-":
            return a - b
        elif operator == "*":
            return a * b
        else:
            return int(a / b)  # Integer division

    def eval_rpn(self, tokens: list[str]) -> int:
        stack = []

        for token in tokens:
            if len(token) == 1 and not token[0].isdigit() and token[0] != "-":
                num2 = stack.pop()
                num1 = stack.pop()
                result = self.resolve(num1, num2, token)
                stack.append(result)
            else:
                stack.append(int(token))

        return stack.pop()


def main():
    calculator = CalculateExpression()
    tokens = ["2", "1", "+", "3", "*"]
    result = calculator.eval_rpn(tokens)
    print(f"Result: {result}")  # Expected: 9


if __name__ == "__main__":
    main()
